solr/core/src/java/org/apache/solr/storage/SizeAwareDirectory.java

```diff
@@ -298,9 +364,17 @@ public long setSizeWriter(LongObjectProcedure<String> sizeWriter) {
     @Override
     @SuppressWarnings("try")
     public void close() throws IOException {
+      long size = backing.getFilePointer();
       try (backing) {
-        fileSizeMap.put(name, backing.getFilePointer());
+        // TODO
       } finally {
+        long onDiskSize = size;
+        if (backing instanceof CompressingDirectory.SizeReportingIndexOutput) {
+          onDiskSize = ((CompressingDirectory.SizeReportingIndexOutput) backing).getBytesWritten();
+        }
+        // we don't know onDiskSize until the file is closed, so write it here
+        sizeWriter.apply(0, onDiskSize, name);
```

Comment:
I suppose if we pull sizeWriter into the DirectIOIndexOutput we could potentially have this updating sooner than close?

Reply:
Without breaking abstraction as much, we could track the last seen `getBytesWritten()` value from within `SizeAccountingIndexOutput`, then update compressed size diff in parallel with updating the uncompressed sizes.

This (or something like) might be a good idea, because reconciliation iirc has a slight (unavoidable, relatively benign) race condition that is made worse by the _actual_ on-disk size (like, as measured from the filesystem) drifting too far away from the in-memory-tracked on-disk size.

Reply:
I've tried reading `getBytesWritten` before calling `writeBytes` and after and using the diff in `sizeWriter`, but that doesn't seem to work. I'm not sure why yet, still looking into that. Unless that's not what you meant?

Reply:
Just kidding, I think I have something working like that now!